Gaussian Mixture - weighted implementation

dupuisIRT opened the pull request:
PR created to work on  #15647

Implementation of a weighted GMM version. If a sample has a weight w, it means the sample is observed w times.

Based on the two papers:
- EM Algorithms for Weighted-Data Clustering with Application to Audio-Visual Scene Analysis, Gebru et al., https://arxiv.org/abs/1509.01509
- Reward-weighted GMM and its application toaction-selection in robotized shoe dressing, Colomé et al., http://www.iri.upc.edu/files/scidoc/1953-Reward-weighted-GMM-and-its-application-to-action-selection-in-robotized-shoe-dressing.pdf

Replicating the structure of Kmeans with a sample_weight argument.
Two main tests are added:
- samples with zero weights do not have an impact
- samples with integer weights should give the same result than a duplicated sample (ex: if a sample X1 has a weight of 3 and all other samples have a unit weight, weighted gmm should give the same result that the classical GMM with three X1 samples + the other samples with no modification)

For the moment, the work is still in progress, in particular regarding:
- PEP8
- the validity of the EM algorithm
- the tests
- the default behavior of sample_weight (if None)

dupuisIRT committed:
Implement weighted-data gmm

- EM updated with sample weight
- Log prob to to the power of sample weight

dupuisIRT committed:
Correct log prob and associated methods

- Instead of (p(x; theta)) ** w, it is Gaussian(x; theta) ** w
- Consistent with https://arxiv.org/pdf/1509.01509.pdf

dupuisIRT committed:
Add tests, update bayesian interface, and tied cov

dupuisIRT committed:
Improve PEP8 and change tests

dupuisIRT committed:
Correct probability value with sampele_weight

dupuisIRT committed:
Correct the working version

dupuisIRT committed:
Merge branch 'master' into gmm_weighted_2

dupuisIRT committed:
Correction to the linting

jnothman commented:
> For the moment, the work is still in progress, in particular regarding ...

Are you looking for feedback at this point? At a glance this is looking quite nice

dupuisIRT committed:
Fix the number of repeated samples in test

- in test_weighted_vs_repeated

dupuisIRT commented:
Hi @jnothman , I think a quick feedback for the test test_weighted_vs_repeated in mixture/tests/test_gaussian_mixture.py is welcome.

jnothman commented:
(We have over time proposed generic testing for sample weight invariances, but it's not really happened outside of evaluation metrics.)

jnothman commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -653,6 +768,71 @@ def test_gaussian_mixture_fit():
+        assert_allclose(g_weighted.means_[arg_idx1],
+                        g_repeated.means_[arg_idx2])
+
+        if covar_type == 'full':
```
I can't say we'd always go to this level of detail in ensuring models are identical.

It would be less repetitive if you defined the transformation once, then applied it to both `precisions_` attributes.

```py
def uniform_precisions(precisions):
    if covar_type == "tied":
        ...
    ...
    return np.trace(precisions, axis1=1, axis2=2).argsort()
```

But I also wonder whether we should just more approximately evaluate that the two models have identical likelihood functions applied to held-out data.

jnothman commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -653,6 +768,71 @@ def test_gaussian_mixture_fit():
+            ecov = EmpiricalCovariance()
+            ecov.covariance_ = prec_weighted[h]
+            # the accuracy depends on the number of data and randomness, rng
+            assert_allclose(ecov.error_norm(prec_repeated[k]), 0, atol=0.001)
```
are we not able to compare precisions directly once indexed by the appropriate `arg_idx`? I.e. can't we just `assert_allclose(g_weighted.precisions_[arg_idx1], g_repeated.precisions_[arg_idx2])` except where 	tied? I don't know this code well, admittedly.

dupuisIRT replied:
Here, I used a conservative approach as the Frobenius error norm was used in other tests, such as test_gaussian_mixture_fit. I changed for `assert_allclose(g_weighted.precisions_[arg_idx1], g_repeated.precisions_[arg_idx2])` and it failed for some tests.

jnothman reviewed (commented):
Some comments!

dupuisIRT commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -653,6 +768,71 @@ def test_gaussian_mixture_fit():
+    n_features = rand_data.n_features
+    n_components = rand_data.n_components
+
+    for covar_type in COVARIANCE_TYPE:
```
I reproduced this structure from other tests (for instance test_gaussian_mixture_fit_best_params).

I generalize @pytest.mark.parametrize to the other tests.

dupuisIRT committed:
Improve test following jnothman's comments

- Add clone to avoid repeating GMM instance
- Change the way to assert weights_
- Add parametrize
- Add uniform_precisions function

dupuisIRT committed:
Correct linting

dupuisIRT commented:
Thanks @jnothman for the comments, I tried to take them in consideration as much as possible.

dupuisIRT commented:
Hey @jnothman , I implemented the majority of your comments. Should I wait for other review or maybe should I try to reach some GMM developers?

cmarmo commented:
Hi @dupuisIRT , I'm sorry your pull request got lost. If you are still interested in finalize this work, do you mind synchronizing with upstream? Please note that in the meanwhile the branch `master` has been renamed as `main`.
Thank you very much for your work so far and your patience.

dupuisIRT commented:
Sure, I'll synchronize with upstream.
